activation/activation.go

```diff
@@ -456,114 +394,118 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
 			return fmt.Errorf("failed to calculate activeset: %v", err)
 		}
 	}
-	if activeSetSize == 0 && !targetEpoch.IsGenesis() {
+	if activeSetSize == 0 && !(pubEpoch + 1).IsGenesis() {
 		return fmt.Errorf("empty active set size found! epochId: %v, len(view): %d, view: %v",
 			pubEpoch, len(view), view)
 	}
 
 	var commitment *types.PostProof
-	if b.prevATX == nil {
+	if b.challenge.PrevATXId == *types.EmptyAtxId {
 		commitment = b.commitment
 	}
 
-	atx := types.NewActivationTxWithChallenge(*b.challenge, b.getCoinbaseAccount(), activeSetSize, view, b.nipst, commitment)
+	atx := types.NewActivationTxWithChallenge(*b.challenge, b.getCoinbaseAccount(), activeSetSize, view, nipst, commitment)
 
-	b.log.With().Info("active ids seen for epoch", log.Uint64("pos_atx_epoch", uint64(posEpoch)),
+	b.log.With().Info("active ids seen for epoch", log.Uint64("atx_pub_epoch", uint64(pubEpoch)),
 		log.Uint32("view_cnt", activeSetSize))
 
-	signedAtx, err := b.SignAtx(atx)
-	if err != nil {
-		return err
-	}
-
-	buf, err := types.InterfaceToBytes(signedAtx)
-	if err != nil {
+	atxReceived := b.db.AwaitAtx(atx.Id())
+	if err := b.signAndBroadcast(atx); err != nil {
 		return err
 	}
-	b.prevATX = atx.ActivationTxHeader
-
-	// cleanup state
-	b.nipst = nil
-	b.challenge = nil
-	b.posLayerID = 0
 
-	err = b.net.Broadcast(AtxProtocol, buf)
-	if err != nil {
-		return err
-	}
+	b.log.Event().Info("atx published!",
+		log.AtxId(atx.ShortString()),
+		log.String("prev_atx_id", atx.PrevATXId.ShortString()),
+		log.String("post_atx_id", atx.PositioningAtx.ShortString()),
+		log.LayerId(uint64(atx.PubLayerIdx)),
+		log.EpochId(uint64(atx.PubLayerIdx.GetEpoch(b.layersPerEpoch))),
+		log.Uint32("active_set", atx.ActiveSetSize),
+		log.String("miner", b.nodeId.Key[:5]),
+		log.Int("view", len(atx.View)),
+	)
 
-	err = b.discardChallenge()
-	if err != nil {
-		log.Error("cannot discard Nipst challenge %s", err)
+	select {
+	case <-atxReceived:
+		b.log.Info("atx received in db")
+		events.Publish(events.AtxCreated{true, atx.ShortString(), uint64(b.currentEpoch())})
+	case <-b.layerClock.AwaitLayer((atx.TargetEpoch(b.layersPerEpoch) + 1).FirstLayer(b.layersPerEpoch)):
+		select {
+		case <-b.syncer.Await():
```

Comment:
why wait to be synced here?

Reply:
We're waiting for the ATX to appear in a block. We're waiting based on clock time, if the timeout expired, it could happen because nobody included our ATX, but it could also happen because we didn't receive the block yet because we're not synced. By waiting to be synced we ensure it's the first case.

There's currently a possibility that someone will include our ATX after the target epoch has ended causing our next ATX to be contextually invalid and us to be blacklisted due to it. #1655 should make this scenario impossible.

Reply:
I've added the comment 
```golang
// ensure we've seen all blocks before concluding that the ATX was lost
```